### PR TITLE
boards: riscv: Add mpfs_icicle board support

### PR DESCRIPTION
**Pull request by con-pax**
Add initial support for the Microchip PolarFire SoC Icicle Kit development Board.

This work adds initial support for 'mpfs_icicle' which has a five core micro-
processor subsystem with:
        SiFive E51 Monitor core (1 x RV64IMAC)
        SiFive U54 Application cores (4 x RV64GC)

The initial support will only run single core applications, on the E51 Monitor
core. 
    
However, SMP support for this board has been proposed in PR #42013.
The aim is to reduce the complexity of PR #42013, by breaking it into smaller PR's

The board also depends on the following PR to add mpfs_hal to the hal_microchip
repository:
https://github.com/zephyrproject-rtos/hal_microchip/pull/15

### CODE REVIEW

File: boards/riscv/mpfs_icicle/Kconfig.board [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2020-2021 Microchip Technology Inc
```
Comment (fkokosinski) [inline boards/riscv/mpfs_icicle/Kconfig.board:1]: Change `2021` to `2022`? This applies also to the other copyright notes.

File: boards/riscv/mpfs_icicle/support/mpfs250t.resc [hunk 2]
```diff
@@ -0,0 +1,17 @@
+:name: Mi-V
+:description: This script is prepared to run Zephyr on a Mi-V RISC-V board.
+
+$name?="Mi-V"
```
Comment (fkokosinski) [inline boards/riscv/mpfs_icicle/support/mpfs250t.resc:4]: `Mi-V` could also refer to the `m2gl025_miv` board. Maybe `mpfs_icicle` would be a better suit here?

File: boards/riscv/mpfs_icicle/support/mpfs250t.resc [hunk 3]
```diff
@@ -0,0 +1,17 @@
+:name: Mi-V
+:description: This script is prepared to run Zephyr on a Mi-V RISC-V board.
```
Comment (fkokosinski) [inline boards/riscv/mpfs_icicle/support/mpfs250t.resc:2]: `Mi-V` could also refer to the `m2gl025_miv` board. Maybe `mpfs_icicle` would be a better suit here?

File: soc/riscv/riscv-privilege/mpfs/soc.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2020-2021 Microchip Technology Inc
+ */
+#ifndef __RISCV64_MPFS_SOC_H_
+#define __RISCV64_MPFS_SOC_H_
+
+#include <soc_common.h>
+#include <devicetree.h>
+
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/soc.h:11]: Unnecessary two blank lines. One should be enough IMO.

File: soc/riscv/riscv-privilege/mpfs/soc.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#define RISCV_MTIME_BASE				0x0200BFF8ULL
+#define RISCV_MTIMECMP_BASE				(0x02004000ULL + (8ULL * 0))
+#define RISCV_MTIMECMP_BY_HART(h)		(0x02004000ULL + (8ULL * (h)))
+#define RISCV_MSIP_BASE					0x02000000
+
+/* lib-c hooks required RAM defined variables */
+#define RISCV_RAM_BASE					CONFIG_SRAM_BASE_ADDRESS
+#define RISCV_RAM_SIZE					KB(CONFIG_SRAM_SIZE)
+
+
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/soc.h:23]: Unnecessary two blank lines. One should be enough IMO.

File: soc/riscv/riscv-privilege/mpfs/linker.ld [hunk 6]
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2018 Antmicro <www.antmicro.com>
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/linker.ld:2]: Wrong copyright notice?

File: soc/riscv/riscv-privilege/mpfs/Kconfig.defconfig.series [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+if SOC_SERIES_RISCV64_MIV
+
+config SOC_SERIES
+	default "mpfs"
+
+# MPFS should be configured so that the mtimer clock is 1MHz independent of the CPU clock...
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 1000000
+#	default 660000
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/mpfs/Kconfig.defconfig.series:13]: Should this commented out line be removed?

**Review by fkokosinski — CHANGES_REQUESTED** [review]
I've wanted to run a simulation of this board on Renode with:
```
west build -p -b mpfs_icicle samples/hello_world -t run
```
But west informs me that simulation is not supported for this platform. I assume that your intention was for this to work (as you've contributed resc scripts in this PR). Could you perhaps take a look at it again to see if everything is set as it should?

File: boards/riscv/mpfs_icicle/board.cmake [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
+
+set(EMU_PLATFORM renode)
```
[the remark this answers is not in the document]
Comment (con-pax) [reply boards/riscv/mpfs_icicle/board.cmake]: @fkokosinski Thank you for taking the time to review this PR, we really appreciate it. I made the suggested changes, you have suggested.

When executing "west build -p -b mpfs_icicle samples/hello_world -t run" I seen the same message as you. By making the change you suggested here, it removed the problem for me. Could you try again please?
Thanks again,

File: boards/riscv/mpfs_icicle/support/mpfs250t.resc [hunk 9]
```diff
@@ -0,0 +1,17 @@
+:name: MPFS-ICICLE-KIT
+:description: This script is prepared to run Zephyr on a PolarFire SoC Icicle Kit RISC-V board.
+
+$name?="MPFS-ICICLE-KIT"
+
+using sysbus
+mach create $name
+machine LoadPlatformDescription @platforms/boards/mpfs-icicle-kit.repl
+
+showAnalyzer uart
```
Comment (fkokosinski) [inline boards/riscv/mpfs_icicle/support/mpfs250t.resc:10]: This should be `mmuart0` I believe.

File: boards/riscv/mpfs_icicle/support/mpfs250t.resc [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+cpu PerformanceInMips 80
```
Comment (fkokosinski) [inline boards/riscv/mpfs_icicle/support/mpfs250t.resc:11]: This should be `e51` I think, as `cpu` is not defined in [polarfire-soc.repl](https://github.com/renode/renode/blob/master/platforms/cpus/polarfire-soc.repl)

**Review by fkokosinski — COMMENTED** [review]
I was able to get the Renode sim running and working correctly by changing these two lines.

**Review by fkokosinski — APPROVED** [review]
Thanks for the changes. LGTM now

**Review by tgorochowik — APPROVED** [review]

**Pull request merged**